programs/bench.c

```diff
@@ -90,440 +87,662 @@ static UTIL_time_t g_displayClock = UTIL_TIME_INITIALIZER;
-        U64 const crcOrig = g_decodeOnly ? 0 : XXH64(srcBuffer, srcSize, 0);
-        UTIL_time_t coolTime;
-        U64 const maxTime = (g_nbSeconds * TIMELOOP_NANOSEC) + 1;
-        U32 nbDecodeLoops = (U32)((100 MB) / (srcSize+1)) + 1;  /* initial conservative speed estimate */
-        U32 nbCompressionLoops = (U32)((2 MB) / (srcSize+1)) + 1;  /* initial conservative speed estimate */
-        U64 totalCTime=0, totalDTime=0;
-        U32 cCompleted=g_decodeOnly, dCompleted=0;
+    {   
+        U64 const crcOrig = (adv->mode == BMK_decodeOnly) ? 0 : XXH64(srcBuffer, srcSize, 0);
 #       define NB_MARKS 4
         const char* const marks[NB_MARKS] = { " |", " /", " =",  "\\" };
         U32 markNb = 0;
-
-        coolTime = UTIL_getTime();
         DISPLAYLEVEL(2, "\r%79s\r", "");
-        while (!cCompleted || !dCompleted) {
 
-            /* overheat protection */
-            if (UTIL_clockSpanMicro(coolTime) > ACTIVEPERIOD_MICROSEC) {
-                DISPLAYLEVEL(2, "\rcooling down ...    \r");
-                UTIL_sleep(COOLPERIOD_SEC);
-                coolTime = UTIL_getTime();
+        if (adv->mode != BMK_decodeOnly) {
+            BMK_initCCtxArgs cctxprep;
+            BMK_customReturn_t compressionResults;
+            int completed = 0;
+            U64 totalLoops = 0, totalTime = 0, fastest = (U64)(-1LL);
+            UTIL_time_t coolTime = UTIL_getTime();
+            cctxprep.ctx = ctx;
+            cctxprep.dictBuffer = dictBuffer;
+            cctxprep.dictBufferSize = dictBufferSize;
+            cctxprep.cLevel = cLevel;
+            cctxprep.comprParams = comprParams;
+            cctxprep.adv = adv;
+            /* Compression */
+            DISPLAYLEVEL(2, "%2s-%-17.17s :%10u ->\r", marks[markNb], displayName, (U32)srcSize);
+            if(adv->loopMode == BMK_timeMode) {
+                U64 maxTime = adv->nbSeconds * TIMELOOP_NANOSEC;
+                unsigned nbLoops = 1;
+                while(!completed) {
+                    /* Overheat protection */
+                    if (UTIL_clockSpanMicro(coolTime) > ACTIVEPERIOD_MICROSEC) {
+                        DEBUGOUTPUT("\rcooling down ...    \r");
+                        UTIL_sleep(COOLPERIOD_SEC);
+                        coolTime = UTIL_getTime();
+                    }
+
+                    compressionResults = BMK_benchFunction(&local_defaultCompress, (void*)ctx, &local_initCCtx, (void*)&cctxprep,
+                    nbBlocks, srcPtrs, srcSizes, cPtrs, cSizes, nbLoops);
+                    if(compressionResults.error) {
+                        results.error = compressionResults.error;
+                        return results;
+                    }
+
+                    {   U64 loopDuration = compressionResults.result.nanoSecPerRun * nbLoops;
+                        totalLoops += nbLoops;
+                        totalTime += loopDuration;
+                        if (loopDuration > 0) { // nanoSec / run
```

Comment:
There is another use case which is interesting : 
whenever `loopDuration` is "too small" (to be defined, typically ~0.5sec if the intended run is one second),
there is a risk of a rounding error, which will impact speed results.
This is especially the case when clock resolution is poor,
which can happen : 
some systems can have clock resolution of ~10ms, so it only progresses in huge steps.
In such a situation, if "real" time is somewhere between 10-19ms, the clock will nonetheless measure only 10ms,
so that's an error of up to a factor 2.

Even at 100ms, such discrepancy rounding error remains significant (up to 10%!).

You want to reduce scope for such error,
hence impose a minimum measurement time for a result to be considered valid.